File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max=0;
    int finddiameter(TreeNode root){
        if(root==null)
            return 0;
        int lh=finddiameter(root.left);
        int rh=finddiameter(root.right);
        if(max<(lh+rh))
            max=lh+rh;
        return Math.max(lh,rh)+1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        finddiameter(root);
        return max;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        global cmax
        cmax = 0
        def traverse(root):
            global cmax
            if not root:
                return 0
            lh = traverse(root.left)
            rh = traverse(root.right)
            cmax = max(cmax, lh+rh)
            return max(lh,rh)+1
        
        traverse(root)
        return cmax